app/controllers/review_mapping_controller.rb

```diff
@@ -51,129 +47,196 @@ def select_metareviewer
     @mapping = ResponseMap.find(params[:id])
   end
 
+
+  #EDITED: If the user is invalid(not present in database), then return with error message.
+  #If reviewer is a participant of the assignment, then only move forward with reviewer assignment.
+  #Work Done: Refactored add_reviewer and extracted some code to make it smaller.
   def add_reviewer
-    assignment = Assignment.find(params[:id])
+    assignment = find_assignment(params[:id])
     topic_id = params[:topic_id]
-    user_id = User.where(name: params[:user][:name]).first.id
-    # If instructor want to assign one student to review his/her own artifact,
+    user_id = if User.where(name: params[:user][:name]).first
+      then User.where(name: params[:user][:name]).first.id
+      else nil
+      end
+    if user_id.nil?
+      flash[:error] = 'The user does not exist.'
+      redirect_to action: 'list_mappings', id: assignment.id, msg: 'The user does not exist.'
+      return
+    end
+    # If instructor wants to assign one student to review his/her own artifact,
     # it should be counted as "self-review" and we need to make /app/views/submitted_content/_selfreview.html.erb work.
-    if TeamsUser.exists?(team_id: params[:contributor_id], user_id: user_id)
+    if is_self_review?(params[:contributor_id], user_id)
       flash[:error] = 'You cannot assign this student to review his/her own artifact.'
+      redirect_to action: 'list_mappings', id: assignment.id
+      return
+    end
+    # Team lazy initialization
+    SignUpSheet.signup_team(assignment.id, user_id, topic_id)
+    msg = ''
+    begin
+      user = User.from_params(params)
+      # Assign the reviewer to the contributor
+      assign_reviewer_to_contributor(user, assignment, params[:contributor_id], topic_id)
+    rescue StandardError => e
+      msg = e.message
+    end
+
+    redirect_to action: 'list_mappings', id: assignment.id, msg: msg
+  end
+  def add_metareviewer
```

Comment:
Since metareviews have not been used since 2011, I wonder why you added code to implement them.  Or, was it just moved from somewhere else?  Where?  Why?